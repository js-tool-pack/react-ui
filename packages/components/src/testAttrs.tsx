import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
