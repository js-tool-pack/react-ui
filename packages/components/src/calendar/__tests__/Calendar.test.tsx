import { getClassNames } from '@tool-pack/basic';
import { render } from '@testing-library/react';
import { testAttrs } from '~/testAttrs';
import { Calendar } from '..';

describe('Calendar', () => {
  testAttrs(Calendar);
  it('basic', () => {
    expect(
      render(
        <Calendar
          today={new Date(2023, 11, 26)}
          value={new Date(2023, 11, 18)}
        />,
      ).container.firstChild,
    ).toMatchSnapshot();
  });
  it('dateCell', () => {
    expect(
      render(
        <Calendar
          dateCell={(date, attrs, status) => (
            <td
              {...attrs}
              className={getClassNames(attrs.className, {
                active: status.isSelected,
              })}
            >
              {date.getDate()}
            </td>
          )}
          today={new Date(2023, 11, 26)}
          value={new Date(2023, 11, 18)}
        />,
      ).container.firstChild,
    ).toMatchSnapshot();
  });
});
