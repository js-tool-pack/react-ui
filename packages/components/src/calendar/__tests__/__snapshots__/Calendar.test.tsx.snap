// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar basic 1`] = `
<div
  class="t-calendar"
>
  <div
    class="t-calendar-header"
  >
    <div
      class="t-calendar-header__month"
    >
      December 2023
    </div>
    <div
      class="t-button-group"
      role="group"
    >
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          <i
            class="t-icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M808.12630286 92.94690423l-81.20590149-81.20590149-500.29730201 500.52713049 500.25899727 500.02916876 81.24420623-81.28251096-419.01479101-418.78496257z"
                fill="currentColor"
              />
            </svg>
          </i>
        </span>
      </button>
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          today
        </span>
      </button>
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          <i
            class="t-icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M636.91389854 512.22982846L217.86080278 92.94690423l81.32081573-81.20590149 500.25899726 500.52713049L299.14331376 1012.25899726l-81.20590149-81.282511z"
                fill="currentColor"
              />
            </svg>
          </i>
        </span>
      </button>
    </div>
  </div>
  <table
    cellpadding="0"
    cellspacing="0"
    class="t-calendar-table"
  >
    <thead>
      <tr>
        <th>
          Sun
        </th>
        <th>
          Mon
        </th>
        <th>
          Tue
        </th>
        <th>
          Wed
        </th>
        <th>
          Thu
        </th>
        <th>
          Fri
        </th>
        <th>
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          26
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          27
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          28
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          29
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          30
        </td>
        <td
          class="t-calendar-cell"
        >
          1
        </td>
        <td
          class="t-calendar-cell"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          3
        </td>
        <td
          class="t-calendar-cell"
        >
          4
        </td>
        <td
          class="t-calendar-cell"
        >
          5
        </td>
        <td
          class="t-calendar-cell"
        >
          6
        </td>
        <td
          class="t-calendar-cell"
        >
          7
        </td>
        <td
          class="t-calendar-cell"
        >
          8
        </td>
        <td
          class="t-calendar-cell"
        >
          9
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          10
        </td>
        <td
          class="t-calendar-cell"
        >
          11
        </td>
        <td
          class="t-calendar-cell"
        >
          12
        </td>
        <td
          class="t-calendar-cell"
        >
          13
        </td>
        <td
          class="t-calendar-cell"
        >
          14
        </td>
        <td
          class="t-calendar-cell"
        >
          15
        </td>
        <td
          class="t-calendar-cell"
        >
          16
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          17
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--active"
        >
          18
        </td>
        <td
          class="t-calendar-cell"
        >
          19
        </td>
        <td
          class="t-calendar-cell"
        >
          20
        </td>
        <td
          class="t-calendar-cell"
        >
          21
        </td>
        <td
          class="t-calendar-cell"
        >
          22
        </td>
        <td
          class="t-calendar-cell"
        >
          23
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          24
        </td>
        <td
          class="t-calendar-cell"
        >
          25
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--today"
        >
          26
        </td>
        <td
          class="t-calendar-cell"
        >
          27
        </td>
        <td
          class="t-calendar-cell"
        >
          28
        </td>
        <td
          class="t-calendar-cell"
        >
          29
        </td>
        <td
          class="t-calendar-cell"
        >
          30
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          31
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          1
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          2
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          3
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          4
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          5
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          6
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Calendar dateCell 1`] = `
<div
  class="t-calendar"
>
  <div
    class="t-calendar-header"
  >
    <div
      class="t-calendar-header__month"
    >
      December 2023
    </div>
    <div
      class="t-button-group"
      role="group"
    >
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          <i
            class="t-icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M808.12630286 92.94690423l-81.20590149-81.20590149-500.29730201 500.52713049 500.25899727 500.02916876 81.24420623-81.28251096-419.01479101-418.78496257z"
                fill="currentColor"
              />
            </svg>
          </i>
        </span>
      </button>
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          today
        </span>
      </button>
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          <i
            class="t-icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M636.91389854 512.22982846L217.86080278 92.94690423l81.32081573-81.20590149 500.25899726 500.52713049L299.14331376 1012.25899726l-81.20590149-81.282511z"
                fill="currentColor"
              />
            </svg>
          </i>
        </span>
      </button>
    </div>
  </div>
  <table
    cellpadding="0"
    cellspacing="0"
    class="t-calendar-table"
  >
    <thead>
      <tr>
        <th>
          Sun
        </th>
        <th>
          Mon
        </th>
        <th>
          Tue
        </th>
        <th>
          Wed
        </th>
        <th>
          Thu
        </th>
        <th>
          Fri
        </th>
        <th>
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          26
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          27
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          28
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          29
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          30
        </td>
        <td
          class="t-calendar-cell"
        >
          1
        </td>
        <td
          class="t-calendar-cell"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          3
        </td>
        <td
          class="t-calendar-cell"
        >
          4
        </td>
        <td
          class="t-calendar-cell"
        >
          5
        </td>
        <td
          class="t-calendar-cell"
        >
          6
        </td>
        <td
          class="t-calendar-cell"
        >
          7
        </td>
        <td
          class="t-calendar-cell"
        >
          8
        </td>
        <td
          class="t-calendar-cell"
        >
          9
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          10
        </td>
        <td
          class="t-calendar-cell"
        >
          11
        </td>
        <td
          class="t-calendar-cell"
        >
          12
        </td>
        <td
          class="t-calendar-cell"
        >
          13
        </td>
        <td
          class="t-calendar-cell"
        >
          14
        </td>
        <td
          class="t-calendar-cell"
        >
          15
        </td>
        <td
          class="t-calendar-cell"
        >
          16
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          17
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--active active"
        >
          18
        </td>
        <td
          class="t-calendar-cell"
        >
          19
        </td>
        <td
          class="t-calendar-cell"
        >
          20
        </td>
        <td
          class="t-calendar-cell"
        >
          21
        </td>
        <td
          class="t-calendar-cell"
        >
          22
        </td>
        <td
          class="t-calendar-cell"
        >
          23
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          24
        </td>
        <td
          class="t-calendar-cell"
        >
          25
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--today"
        >
          26
        </td>
        <td
          class="t-calendar-cell"
        >
          27
        </td>
        <td
          class="t-calendar-cell"
        >
          28
        </td>
        <td
          class="t-calendar-cell"
        >
          29
        </td>
        <td
          class="t-calendar-cell"
        >
          30
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          31
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          1
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          2
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          3
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          4
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          5
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          6
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Calendar 默认不选中今天 1`] = `
<div
  class="t-calendar"
>
  <div
    class="t-calendar-header"
  >
    <div
      class="t-calendar-header__month"
    >
      December 2023
    </div>
    <div
      class="t-button-group"
      role="group"
    >
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          <i
            class="t-icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M808.12630286 92.94690423l-81.20590149-81.20590149-500.29730201 500.52713049 500.25899727 500.02916876 81.24420623-81.28251096-419.01479101-418.78496257z"
                fill="currentColor"
              />
            </svg>
          </i>
        </span>
      </button>
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          today
        </span>
      </button>
      <button
        class="t-button t-button--type-default t-button--icon-l t--size-sm"
      >
        <span>
          <i
            class="t-icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M636.91389854 512.22982846L217.86080278 92.94690423l81.32081573-81.20590149 500.25899726 500.52713049L299.14331376 1012.25899726l-81.20590149-81.282511z"
                fill="currentColor"
              />
            </svg>
          </i>
        </span>
      </button>
    </div>
  </div>
  <table
    cellpadding="0"
    cellspacing="0"
    class="t-calendar-table"
  >
    <thead>
      <tr>
        <th>
          Sun
        </th>
        <th>
          Mon
        </th>
        <th>
          Tue
        </th>
        <th>
          Wed
        </th>
        <th>
          Thu
        </th>
        <th>
          Fri
        </th>
        <th>
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          26
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          27
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          28
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          29
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--prev-month"
        >
          30
        </td>
        <td
          class="t-calendar-cell"
        >
          1
        </td>
        <td
          class="t-calendar-cell"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          3
        </td>
        <td
          class="t-calendar-cell"
        >
          4
        </td>
        <td
          class="t-calendar-cell"
        >
          5
        </td>
        <td
          class="t-calendar-cell"
        >
          6
        </td>
        <td
          class="t-calendar-cell"
        >
          7
        </td>
        <td
          class="t-calendar-cell"
        >
          8
        </td>
        <td
          class="t-calendar-cell"
        >
          9
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          10
        </td>
        <td
          class="t-calendar-cell"
        >
          11
        </td>
        <td
          class="t-calendar-cell"
        >
          12
        </td>
        <td
          class="t-calendar-cell"
        >
          13
        </td>
        <td
          class="t-calendar-cell"
        >
          14
        </td>
        <td
          class="t-calendar-cell"
        >
          15
        </td>
        <td
          class="t-calendar-cell"
        >
          16
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          17
        </td>
        <td
          class="t-calendar-cell"
        >
          18
        </td>
        <td
          class="t-calendar-cell"
        >
          19
        </td>
        <td
          class="t-calendar-cell"
        >
          20
        </td>
        <td
          class="t-calendar-cell"
        >
          21
        </td>
        <td
          class="t-calendar-cell"
        >
          22
        </td>
        <td
          class="t-calendar-cell"
        >
          23
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          24
        </td>
        <td
          class="t-calendar-cell"
        >
          25
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--today"
        >
          26
        </td>
        <td
          class="t-calendar-cell"
        >
          27
        </td>
        <td
          class="t-calendar-cell"
        >
          28
        </td>
        <td
          class="t-calendar-cell"
        >
          29
        </td>
        <td
          class="t-calendar-cell"
        >
          30
        </td>
      </tr>
      <tr>
        <td
          class="t-calendar-cell"
        >
          31
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          1
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          2
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          3
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          4
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          5
        </td>
        <td
          class="t-calendar-cell t-calendar-cell--next-month"
        >
          6
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
